cnd/src/libp2p_bam/handler.rs

```diff
@@ -43,52 +47,80 @@ pub struct BamHandler<TSubstream> {
 
 #[derive(strum_macros::Display)]
 #[allow(missing_debug_implementations)]
-/// State of an active substream, opened either by us or by the remote.
-enum SubstreamState<TSubstream> {
+/// State of an active substream opened by us.
+enum OutSubstreamState<TSubstream> {
     /// We haven't started opening the outgoing substream yet.
-    OutWaitingOpen { req: PendingOutgoingRequest },
+    WaitingOpen { req: PendingOutgoingRequest },
     /// Waiting to send a message to the remote.
-    OutWaitingSend {
+    WaitingSend {
         msg: Frame,
         response_sender: oneshot::Sender<Response>,
         stream: BamStream<TSubstream>,
     },
     /// Waiting to flush the substream so that the data arrives to the remote.
-    OutWaitingFlush {
+    WaitingFlush {
         response_sender: oneshot::Sender<Response>,
         stream: BamStream<TSubstream>,
     },
     /// Waiting for the answer to our message
-    OutWaitingAnswer {
+    WaitingAnswer {
         response_sender: oneshot::Sender<Response>,
         stream: BamStream<TSubstream>,
     },
+    /// The substream is being closed.
+    WaitingClose { stream: BamStream<TSubstream> },
+}
+
+impl<TSubstream> CloseStream for OutSubstreamState<TSubstream> {
+    type TSubstream = TSubstream;
+
+    fn close(stream: BamStream<Self::TSubstream>) -> Self {
+        OutSubstreamState::WaitingClose { stream }
+    }
+}
+
+#[derive(strum_macros::Display)]
+#[allow(missing_debug_implementations)]
+/// State of an active substream opened by peer node.
+enum InSubstreamState<TSubstream> {
     /// Waiting for a request from the remote.
-    InWaitingMessage { stream: BamStream<TSubstream> },
+    WaitingMessage { stream: BamStream<TSubstream> },
     /// Waiting for the user to send the response back to us.
-    InWaitingUser {
+    WaitingUser {
         response_receiver: oneshot::Receiver<Response>,
         stream: BamStream<TSubstream>,
     },
     /// Waiting to send an answer back to the remote.
-    InWaitingSend {
+    WaitingSend {
```

Comment:
I guess I miss the libp2p knowledge to under the difference between `send` and `flush`.